### ISSUE TITLE
Directory `owner:group` changing back to default user (unknown root cause)

### ISSUE THREAD

**Issue by DeadEnded** [ogarcia/docker-supysonic#2]
Hello again... I was digging into an issue I have been seeing where my container is having errors after a recreation.
I finally noticed that the directory and all files in it that are mapped to the container volumes are resetting the user and group from the 100:100 to 1000:1000.

I found this very odd.. basically going to the default user... wasn't sure what was causing this - but then I saw on line 14 of the dockerfile `  chown supysonic:users /var/lib/supysonic /var/log/supysonic && \`.  I believe this might be an issue because ` supysonic:users` doesn't exist on my host.  My guess is that since it cannot find the user defined in the dockerfile - it is using the default (or primary whatever you would call it) user.

To fix this, do I need to add a line to my docker run/compose command to run as user 100? `--user 100:100`
Or is there something else going on?

Thanks!
DeadEnd

**Comment by ogarcia** [comment]
> I found this very odd.. basically going to the default user... wasn't sure what was causing this - but then I saw on line 14 of the dockerfile ` chown supysonic:users /var/lib/supysonic /var/log/supysonic && \`. I believe this might be an issue because ` supysonic:users` doesn't exist on my host. My guess is that since it cannot find the user defined in the dockerfile - it is using the default (or primary whatever you would call it) user.

The sentence `chown supysonic:users /var/lib/supysonic /var/log/supysonic` is executed **inside** of container where the user _supysonic_ and group _users_ exist. This is neccesary if you simply want to run image to make a demo or test supysonic,

> To fix this, do I need to add a line to my docker run/compose command to run as user 100? `--user 100:100`
> Or is there something else going on?

Yes, your problem can be in one hand that you are mounting a volume from your host machine into container. The directory in your host must be permissions before use it. In this case you can do:
```
mkdir -p /my/supysonic/data /my/supysonic/logs
chown -R 100:100 /my/supysonic/data /my/supysonic/log
docker run -d \
  --name=supysonic \
  -v /my/supysonic/data:/var/lib/supysonic \
  -v /my/supysonic/logs:/var/log/supysonic \
  -v /my/media/directory:/media \
  ogarcia/supysonic
```

In other hand your problem can be that you are using a docker volume. In this case you must run the `chown -R 100:100 /my/supysonic/data /my/supysonic/log` before start the supysonic container. In kubernetes you can do this with a [init container](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/). With `docker-compose` maybe using [`depends_on`](https://docs.docker.com/compose/compose-file/compose-file-v3/#depends_on), I don`t know, I don't use compose :shrug:

**Comment by DeadEnded** [comment]
yes, each time it resets (I believe on recreation of the container) I have to run `chown -R 100:100 /my/supysonic/data /my/supysonic/log` on my mapped volumes on the host.  I do this after the container starts because that is when it breaks.

....

Okay, I guess I lied... I just tried both the variations (stop/start and recreate) and the volumes didn't change back to default user... so something is causing it but not that...  So something has caused the directories to revert back to the default user as owner and group... but I don't know what...

I need to go back and do more testing... I thought I knew what was causing it - but apparently, I was wrong.  Sorry for the run-around!

Well, I'm going to close this issue for now and re-open it when I identify what the cause is (if I still can't figure out the fix).
If I figure it out and have a fix, I will just comment on the solution without re-opening. (so if anyone has it in the future it's documented)
Sorry again for opening this without making sure I had the true source of the problem!

Cheers!
DeadEnd

**Comment by DeadEnded** [comment]
I know it has been some time... I still have this issue, and recently decided I needed to find out why.
So after some digging I found out the cause of this is when the docker daemon starts... all my bind mounts have their owner and group changed to the default (1000:1000).  The reason I haven't noticed this before is because the only containers that don't already have 1000:1000 set are postgres database containers.  I checked out their entrypoint.sh and sure enough they fix this on their own.  I believe the below code taken from their entrypoint is where they fix it ( I have watch the directory during docker daemon start and see it change to 1000 and back to 70 which is the correct user for postgres):

```
# used to create initial postgres directories and if run as root, ensure ownership to the "postgres" user
docker_create_db_directories() {
        local user; user="$(id -u)"

        mkdir -p "$PGDATA"
        # ignore failure since there are cases where we can't chmod (and PostgreSQL might fail later anyhow - it's picky about permissions of this directory)
        chmod 700 "$PGDATA" || :

        # ignore failure since it will be fine when using the image provided directory; see also https://github.com/docker-library/postgres/pull/289
        mkdir -p /var/run/postgresql || :
        chmod 775 /var/run/postgresql || :

        # Create the transaction log directory before initdb is run so the directory is owned by the correct user
        if [ -n "$POSTGRES_INITDB_WALDIR" ]; then
                mkdir -p "$POSTGRES_INITDB_WALDIR"
                if [ "$user" = '0' ]; then
                        find "$POSTGRES_INITDB_WALDIR" \! -user postgres -exec chown postgres '{}' +
                fi
                chmod 700 "$POSTGRES_INITDB_WALDIR"
        fi

        # allow the container to be started with `--user`
        if [ "$user" = '0' ]; then
                find "$PGDATA" \! -user postgres -exec chown postgres '{}' +
                find /var/run/postgresql \! -user postgres -exec chown postgres '{}' +
        fi
}
```

So it appears the source of my problem is Docker itself...
I'm not sure why other users haven't had this issue... perhaps they aren't using bind mounts for their config folder.
I'm am going to keep digging to see if I can find a solution to prevent docker from doing this.
Otherwise I will have to write up something based on what I see in the postgres entrypoint to force the correct user on startup.

Cheers and thanks for listening to my rambling!
DeadEnd

**Comment by ogarcia** [comment]
This is a strange behaviour. A directory must not change his ownership automagically. Maybe something related with SELinux? The directories are stored in your home?

Other thing that you can try is leave docker manage the volumes:
```sh
% docker volume create supysonic-data
% docker volume create supysonic-logs
% docker volume inspect supysonic-data # <-- To kown where is stored
[
    {
        "CreatedAt": "2021-11-23T09:19:32+01:00",
        "Driver": "local",
        "Labels": {},
        "Mountpoint": "/var/lib/docker/volumes/supysonic-data/_data", <- Here
        "Name": "supysonic-data",
        "Options": {},
        "Scope": "local"
    }
]
% chown -R 100:100 /var/lib/docker/volumes/supysonic-data/_data
% chown -R 100:100 /var/lib/docker/volumes/supysonic-logs/_data
```

**Comment by DeadEnded** [comment]
Apologies for the log delay...
No, the directory is in a custom location based on my crazy thoughts...
```
/docker/supysonic/
```
My reason for this is all my configurations are in sub-directories of the `/docker/` directory.  I back up this one directory and have a backup of all configurations.  Anyway, that is the method to my madness.  As I said, I was refreshing the `/docker/` directory repeatedly and that's when I discovered the behavior.  All sub-directories in `/docker/` changed their user and group ownership... and then the postgres sub-directory fixed itself a few seconds later (which led me to investigate and find the above code that appears to fix it).

I haven't had the time recently to research and see how common this is and if there are other solutions to fix it.

Anyway, thank you for your time - I appreciate your help understanding this!  I think you solution of having docker manage the image is a likely solution - but I like having my config directories mapped to the host... so I'll try and find a solution to this and let you know if/when I do!

Cheers!
DeadEnd

**Comment by ogarcia** [comment]
If you leave Docker manage your volumes then you only have to backup `/var/lib/docker/volumes/`. But is very similar to have all volumes in `/docker`.

Maybe your backup script are doing the permissions change?

**Comment by DeadEnded** [comment]
I don't backup all volumes - only configuration volumes.
I have other volumes on external storage for example for large file storage, etc.

But to your point - that absolutely would work - just have to configure it properly.
As far as what is causing it - I am nearly 100% sure it is docker daemon.
Last month when I resurrected this discussion I was able to watch it change (both the config folder for supysonic and postgres) and then see postgress "fix" itself directly after issuing a restart of docker daemon.  I did it multiple times in a row with 100% reproducibility.

The fact that postgres has a script to fix this makes me think it is a docker bug... but I am going to try and get some others to confirm if they can reproduce it as well.  I need to do more research (life has been very busy) to see if there is a solution out there already.

DeadEnd
